### PR TITLE
Enable contact search in Email skill

### PR DESCRIPTION
**Pull request by DingmaomaoBJTU**
## Description
Enable contact search in Email skill

## Related Issue
Cannot email an specific contact #248

## Testing Steps
1. In your MSA account, add an new contact named "Test" and set email as "test@outlook.com"(Ensure this address never send you an email, otherwise it will be include in people you work with)
2. Open email bot -> "send an email to Test"
3. Test is resolved as contact.

### CODE REVIEW

File: solutions/Virtual-Assistant/src/csharp/skills/emailskill/Dialogs/Shared/EmailSkillDialog.cs [hunk 1]
```diff
@@ -453,17 +453,22 @@ public async Task<bool> ChoiceValidator(PromptValidatorContext<FoundChoice> pc,
         }
 
         // Helpers
-        public async Task<string> GetNameListString(WaterfallStepContext sc)
+        public async Task<string> GetNameListStringAsync(WaterfallStepContext sc)
```
Comment (lzc850612) [inline solutions/Virtual-Assistant/src/csharp/skills/emailskill/Dialogs/Shared/EmailSkillDialog.cs]: does this function have to be public? #Resolved

File: solutions/Virtual-Assistant/src/csharp/skills/emailskill/Dialogs/Shared/EmailSkillDialog.cs [hunk 2]
```diff
@@ -453,17 +453,22 @@ public async Task<bool> ChoiceValidator(PromptValidatorContext<FoundChoice> pc,
...
         {
-            var state = await _emailStateAccessor.GetAsync(sc.Context);
+            var state = await _emailStateAccessor.GetAsync(sc?.Context);
             var recipients = state.Recipients;
-            if (recipients.Count == 1)
+
+            if (recipients.Count == 0)
+            {
+                throw new Exception("Invalid user number!");
```
Comment (lzc850612) [inline solutions/Virtual-Assistant/src/csharp/skills/emailskill/Dialogs/Shared/EmailSkillDialog.cs]: No contact or No recipient? Invalid user number feels weird #Resolved

File: solutions/Virtual-Assistant/src/csharp/skills/emailskill/Dialogs/Shared/EmailSkillDialog.cs [hunk 3]
```diff
@@ -478,6 +483,73 @@ public async Task<string> GetNameListString(WaterfallStepContext sc)
...
+            foreach (var person in personList)
+            {
+                var mailAddress = person.ScoredEmailAddresses.FirstOrDefault()?.Address ?? person.UserPrincipalName;
+
+                bool isDup = false;
+                foreach (var formattedPerson in formattedPersonList)
+                {
+                    var formattedMailAddress = formattedPerson.ScoredEmailAddresses.FirstOrDefault()?.Address ?? formattedPerson.UserPrincipalName;
+
+                    if (mailAddress.Equals(formattedMailAddress))
```
Comment (lzc850612) [inline solutions/Virtual-Assistant/src/csharp/skills/emailskill/Dialogs/Shared/EmailSkillDialog.cs]: add option, specifically ignorecase #Resolved

File: solutions/Virtual-Assistant/src/csharp/skills/emailskill/Dialogs/Shared/EmailSkillDialog.cs [hunk 4]
```diff
@@ -770,6 +842,30 @@ public async Task<List<Person>> GetUserAsync(WaterfallStepContext sc, string nam
...
+                // Get users.
+                var contactsList = await service.GetContactsAsync(name);
+                foreach (var contact in contactsList)
+                {
+                    result.Add(contact.ToPerson());
+                }
+            }
+            catch (ServiceException)
+            {
+                // won't clear conversation state hear, because sometime use api is not available, like user msa account.
```
Comment (lzc850612) [inline solutions/Virtual-Assistant/src/csharp/skills/emailskill/Dialogs/Shared/EmailSkillDialog.cs]: we just swallow the exception and do nothing? we should at least throw the exception back right? #Resolved

File: solutions/Virtual-Assistant/src/csharp/skills/emailskill/Extensions/UserEx.cs [hunk 5]
```diff
@@ -35,5 +35,27 @@ public static Person ToPerson(this User user)
...
+            {
+                DisplayName = contact.DisplayName,
+                UserPrincipalName = null,
+                Surname = contact.Surname,
+            };
+            var emailAddresses = new List<ScoredEmailAddress>
+            {
+                new ScoredEmailAddress()
+                {
+                    Address = contact.EmailAddresses?.FirstOrDefault()?.Address,
```
Comment (lzc850612) [inline solutions/Virtual-Assistant/src/csharp/skills/emailskill/Extensions/UserEx.cs:61]: check contact before using it #Resolved

**Review by lzc850612 — CHANGES_REQUESTED** [review]
left comments

File: solutions/Virtual-Assistant/src/csharp/skills/emailskill/Dialogs/Shared/EmailSkillDialog.cs (hunk 1 above)
Comment (DingmaomaoBJTU) [reply solutions/Virtual-Assistant/src/csharp/skills/emailskill/Dialogs/Shared/EmailSkillDialog.cs]: change to protected. Public was for the UT but ut can work with inherit. Change ut accordingly. #Resolved

File: solutions/Virtual-Assistant/src/csharp/skills/emailskill/Dialogs/Shared/EmailSkillDialog.cs (hunk 2 above)
Comment (DingmaomaoBJTU) [reply solutions/Virtual-Assistant/src/csharp/skills/emailskill/Dialogs/Shared/EmailSkillDialog.cs]: Change to no recipient~ thanks! #Resolved

File: solutions/Virtual-Assistant/src/csharp/skills/emailskill/Extensions/UserEx.cs (hunk 5 above)
Comment (DingmaomaoBJTU) [reply solutions/Virtual-Assistant/src/csharp/skills/emailskill/Extensions/UserEx.cs:61]: Good suggestion, thanks! #Resolved

File: solutions/Virtual-Assistant/src/csharp/skills/emailskill/Dialogs/Shared/EmailSkillDialog.cs (hunk 4 above)
Comment (DingmaomaoBJTU) [reply solutions/Virtual-Assistant/src/csharp/skills/emailskill/Dialogs/Shared/EmailSkillDialog.cs]: for get people and get contacts, the exception should be throwed back. But for domain user search, the exception appears when using msa account. I didn't find a way to detect the msa account and aad account, do you have any suggestion on this part? Maybe we should just log the exception here? #Resolved

File: solutions/Virtual-Assistant/src/csharp/skills/emailskill/Dialogs/Shared/EmailSkillDialog.cs (hunk 3 above)
Comment (DingmaomaoBJTU) [reply solutions/Virtual-Assistant/src/csharp/skills/emailskill/Dialogs/Shared/EmailSkillDialog.cs]: Sorry, I didn't get what you mean... could you explain more? #Resolved

Comment (lauren-mills) [reply solutions/Virtual-Assistant/src/csharp/skills/emailskill/Dialogs/Shared/EmailSkillDialog.cs]: I think he means something like this: `mailAddress.Equals(formattedMailAddress,  StringComparison.OrdinalIgnoreCase)` 😊 #Resolved

Comment (DingmaomaoBJTU) [reply solutions/Virtual-Assistant/src/csharp/skills/emailskill/Dialogs/Shared/EmailSkillDialog.cs]: Ah! Got it, thank you so much! 😄 #Resolved

File: solutions/Virtual-Assistant/src/csharp/skills/emailskill/Dialogs/Shared/EmailSkillDialog.cs [hunk 6]
```diff
@@ -454,17 +460,22 @@ protected async Task<bool> ChoiceValidator(PromptValidatorContext<FoundChoice> p
...
         // Helpers
-        protected async Task<string> GetNameListString(WaterfallStepContext sc)
+        protected async Task<string> GetNameListStringAsync(WaterfallStepContext sc)
         {
-            var state = await EmailStateAccessor.GetAsync(sc.Context);
+            var state = await EmailStateAccessor.GetAsync(sc?.Context);
             var recipients = state.Recipients;
-            if (recipients.Count == 1)
+
+            if (recipients.Count == 0)
```
Comment (lzc850612) [inline solutions/Virtual-Assistant/src/csharp/skills/emailskill/Dialogs/Shared/EmailSkillDialog.cs]: check if recipients is null as well #Resolved

File: solutions/Virtual-Assistant/src/csharp/skills/emailskill/ServiceClients/IUserService.cs [hunk 7]
```diff
@@ -22,5 +22,12 @@ public interface IUserService
         /// <param name="name">The person's name.</param>
         /// <returns>A <see cref="Task{TResult}"/> representing the result of the asynchronous operation.</returns>
         Task<List<User>> GetUserAsync(string name);
+
+        /// <summary>
+        /// Get user from your organization.
```
Comment (lzc850612) [inline solutions/Virtual-Assistant/src/csharp/skills/emailskill/ServiceClients/IUserService.cs]: Get contact from your organization #Resolved

File: solutions/Virtual-Assistant/src/csharp/skills/tests/emailskilltest/API/Fakes/MockDialogStateAccessor.cs [hunk 8]
```diff
@@ -0,0 +1,40 @@
+﻿using Microsoft.Bot.Builder;
+using Microsoft.Bot.Builder.Dialogs;
+using Moq;
+using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
```
Comment (lzc850612) [inline solutions/Virtual-Assistant/src/csharp/skills/tests/emailskilltest/API/Fakes/MockDialogStateAccessor.cs]: system usings come first. please pay attention to the stylecop warnings. #Resolved

Comment (lzc850612) [reply solutions/Virtual-Assistant/src/csharp/skills/tests/emailskilltest/API/Fakes/MockDialogStateAccessor.cs]: Please check all the files in this change for this #Resolved

File: solutions/Virtual-Assistant/src/csharp/skills/tests/emailskilltest/API/Helper/StepHelperTests.cs [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+
+            Assert.AreEqual(personList.Count, 6);
+            Assert.AreEqual(userList.Count, 1);
+        }
+
+        private List<Recipient> GetRecipients(int count)
+        {
+            List<Recipient> result = new List<Recipient>();
+
+            for (int i = 0; i < count; ++i)
```
Comment (lzc850612) [inline solutions/Virtual-Assistant/src/csharp/skills/tests/emailskilltest/API/Helper/StepHelperTests.cs]: i++ #Resolved

File: solutions/Virtual-Assistant/src/csharp/skills/tests/emailskilltest/API/Helper/StepHelperTests.cs [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+            }
+
+            return result;
+        }
+
+        private List<Person> GetPersonLists(int start, int end)
+        {
+            List<Person> result = new List<Person>();
+
+            for (int i = start; i < end; ++i)
```
Comment (lzc850612) [inline solutions/Virtual-Assistant/src/csharp/skills/tests/emailskilltest/API/Helper/StepHelperTests.cs]: i++ #Resolved

File: solutions/Virtual-Assistant/src/csharp/skills/tests/emailskilltest/API/Service/UserServiceTests.cs [hunk 11]
```diff
@@ -101,5 +101,48 @@ public async Task GetPeopleTest()
...
+            var mockGraphServiceClient = new MockGraphServiceClientGen();
+            mockGraphServiceClient.People = people;
+            mockGraphServiceClient.SetMockBehavior();
+
+            IGraphServiceClient serviceClient = mockGraphServiceClient.GetMockGraphServiceClient().Object;
+            UserService userService = new UserService(serviceClient, timeZoneInfo: TimeZoneInfo.Local);
+
+            var result = await userService.GetPeopleAsync("test");
+
+            // Test get > 0 people per page
```
Comment (lzc850612) [inline solutions/Virtual-Assistant/src/csharp/skills/tests/emailskilltest/API/Service/UserServiceTests.cs]: > 0 ? #Resolved

**Review by lzc850612 — CHANGES_REQUESTED** [review]

File: solutions/Virtual-Assistant/src/csharp/skills/tests/emailskilltest/API/Helper/StepHelperTests.cs (hunk 9 above)
Comment (DingmaomaoBJTU) [reply solutions/Virtual-Assistant/src/csharp/skills/tests/emailskilltest/API/Helper/StepHelperTests.cs]: Thanks! Fixed. This also a C++ style code as it is said ++i is faster in C++.

---
In reply to: [233171739](https://github.com/Microsoft/AI/pull/249#discussion_r233171739) [](ancestors = 233171739)

File: solutions/Virtual-Assistant/src/csharp/skills/tests/emailskilltest/API/Fakes/MockDialogStateAccessor.cs (hunk 8 above)
Comment (DingmaomaoBJTU) [reply solutions/Virtual-Assistant/src/csharp/skills/tests/emailskilltest/API/Fakes/MockDialogStateAccessor.cs]: Sure!

---
In reply to: [233171096](https://github.com/Microsoft/AI/pull/249#discussion_r233171096) [](ancestors = 233171096)

File: solutions/Virtual-Assistant/src/csharp/skills/tests/emailskilltest/API/Service/UserServiceTests.cs (hunk 11 above)
Comment (DingmaomaoBJTU) [reply solutions/Virtual-Assistant/src/csharp/skills/tests/emailskilltest/API/Service/UserServiceTests.cs]: Remove invalid comments

---
In reply to: [233172225](https://github.com/Microsoft/AI/pull/249#discussion_r233172225) [](ancestors = 233172225)

**Review by lzc850612 — APPROVED** [review]

**Pull request merged**